### PR TITLE
Added Coding Guidance to the Developer Guide documentation.

### PR DESCRIPTION
**Pull request by dlvenable**
### Description

Adds coding guidance for the project into our existing Developer Documentation page.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: docs/developer_guide.md [hunk 1]
```diff
@@ -61,6 +61,83 @@ java -jar data-prepper-core-$VERSION.jar pipelines.yaml data-prepper-config.yaml
...
+2. Avoid abbreviations unless they are widely accepted
+3. Use final on all variables which are not reassigned
+4. Wildcard imports are not allowed.
+5. Static imports are preferred over qualified imports when using static methods
+6. Prefer creating non-static methods whenever possible. Static methods should generally be avoid as they are often used as a shortcut. Sometimes static methods are the best solution such as when using a builder.
+7. Public utility or “common” classes are not permitted.
+    1. They are fine in test code
+    2. They are fine if package protected
+8. Use Optional for return values if the value may not be present. This should be preferred to returning null.
+9. Do not create checked exceptions, and do not throw checked exceptions whenever possible. In general, if you call a method with a checked exception, you should wrap that exception into an unchecked exception.
```
Comment (chenqi0805) [inline docs/developer_guide.md:89]: Should this only apply to APIs meant to be overridden? For internal/private helper method, I am not sure if we have added too much restriction on code style here.

File: docs/developer_guide.md [hunk 2]
```diff
@@ -61,6 +61,83 @@ java -jar data-prepper-core-$VERSION.jar pipelines.yaml data-prepper-config.yaml
...
+3. Use Mockito for mocking
+4. Each class should have a unit test.
+5. Unit test class names should end with Test.
+6. Each large component should have an integration test.
+   1. A good example is a plugin. Plugins should have their own integration tests which integrate all of the plugin’s classes. However, these tests do not run a full Data Prepper.
+7. Integration test class names should end with IT.
+8. Test names should indicate what is being tested, if we see a failed test we should be able to look at the test name and have a good idea about what just failed with minimal context about the code being written
+   1. Two good approaches may be used, depending on what you are testing:
+      1. methodUnderTest_condition_result
+      2. test_when_something_condition_then_something_else
```
Comment (chenqi0805) [inline docs/developer_guide.md:133]: 👍

**Review by sshivanii — DISMISSED** [review]
LGTM

File: docs/developer_guide.md (hunk 1 above)
Comment (dlvenable) [reply docs/developer_guide.md:89]: Yes, that is my intention, and I can clarify it.

Comment (dlvenable) [reply docs/developer_guide.md:89]: How is this?

> 9. Do not create checked exceptions, and do not throw checked exceptions from public methods whenever possible. In general, if you call a method with a checked exception, you should wrap that exception into an unchecked exception.
    1. Throwing checked exceptions from private methods is acceptable.

Comment (chenqi0805) [reply docs/developer_guide.md:89]: Works for me.

**Review by chenqi0805 — APPROVED** [review]

**Review by sshivanii — APPROVED** [review]

**Pull request merged**